File: data-science-major/labs/course-6-r/python/01_descriptive.py
"""Experiment 1 (Python equivalent) -- mean, median, mode, variance, SD.

R version: ../01_descriptive.R
The R script quotes these numbers; this file is what verifies them.
"""

import statistics
from collections import Counter

from _shared import MARKS


def describe(values):
    n = len(values)
    mean = sum(values) / n
    ordered = sorted(values)
    median = (statistics.median(values))
    counts = Counter(values)
    top = max(counts.values())
    modes = sorted(v for v, c in counts.items() if c == top)

    # R's var() and sd() are the SAMPLE versions -- divide by n-1.
    sample_var = sum((x - mean) ** 2 for x in values) / (n - 1)
    pop_var = sum((x - mean) ** 2 for x in values) / n

    return {
        "n": n, "mean": mean, "median": median,
        "mode": modes if top > 1 else None,
        "sample_var": sample_var, "sample_sd": sample_var ** 0.5,
        "pop_var": pop_var, "pop_sd": pop_var ** 0.5,
        "range": max(values) - min(values),
    }


if __name__ == "__main__":
    r = describe(MARKS)
    print("EXPERIMENT 1 -- Descriptive statistics")
    print(f"  n            = {r['n']}")
    print(f"  mean         = {r['mean']:.4f}")
    print(f"  median       = {r['median']:.4f}")
    print(f"  mode         = {r['mode'] if r['mode'] else 'none (all values unique)'}")
    print(f"  range        = {r['range']}")
    print(f"  var  (n-1)   = {r['sample_var']:.4f}   <- R's var()")
    print(f"  sd   (n-1)   = {r['sample_sd']:.4f}   <- R's sd()")
    print(f"  var  (n)     = {r['pop_var']:.4f}   <- population")
    print(f"  sd   (n)     = {r['pop_sd']:.4f}")

    assert abs(r["mean"] - statistics.mean(MARKS)) < 1e-9
    assert abs(r["sample_var"] - statistics.variance(MARKS)) < 1e-9
    assert abs(r["pop_var"] - statistics.pvariance(MARKS)) < 1e-9
    print("\n  cross-checked against the statistics module ✓")
    print("\n  NOTE: R's var() and sd() use n-1. R has no built-in mode();")
    print("        the R script defines one, as the syllabus expects.")

File: data-science-major/labs/course-6-r/01_descriptive.R
# =====================================================================
# NOT EXECUTED IN VERIFICATION
#
# R could not be installed in the environment where this repository was
# verified -- the Debian package repositories are blocked by the network
# policy, and R is not available from PyPI or npm. This script is
# DESK-CHECKED ONLY: the syntax has been reviewed by hand, not run.
#
# The numbers quoted in the comments below come from the Python equivalent
# in python/, which WAS executed. So the statistics are machine-checked even
# though the R syntax is not. Run this in RStudio to confirm the syntax.
# =====================================================================
# Experiment 1: Mean, Median, Mode, Variance, Standard Deviation
# Python equivalent: python/01_descriptive.py

marks <- c(45, 67, 78, 52, 89, 91, 73, 64, 58, 82,
           76, 69, 71, 85, 60, 55, 93, 48, 79, 66)

# --- Central tendency ---
mean(marks)      # 70.05
median(marks)    # 70.00

# R has NO built-in mode() for the statistical mode -- mode() reports the
# storage type. Define one, which the syllabus expects you to know:
statistical_mode <- function(v) {
  freq <- table(v)
  as.numeric(names(freq)[freq == max(freq)])
}
statistical_mode(marks)   # every value occurs once here, so all 20 are returned

# --- Dispersion ---
var(marks)       # 202.8921   <- R's var() divides by n-1 (SAMPLE)
sd(marks)        #  14.2440   <- likewise
range(marks)     # 45 93
diff(range(marks))  # 48
IQR(marks)
quantile(marks)

# Population variance, if you need it, must be computed explicitly:
n <- length(marks)
var(marks) * (n - 1) / n     # 192.7475
sqrt(var(marks) * (n - 1) / n)   # 13.8834

# --- Everything at once ---
summary(marks)

# NOTE FOR THE EXAM: R's var() and sd() are the n-1 versions. If a question
# asks for the population variance you must convert, as above. This is the
# same n vs n-1 distinction as Course 4 Unit 1.
